Exit on failed sanity check, provide config to skip exit

RolandStuder opened the pull request:
# Enhancement of sanity checks

## Description

This PR again exits on failed sanity checks, but provides a config option to not do so. @excid3 provided us with wonderful sanity checks on startup, as to reduce confusion when caching is not enabled or there is a version mismatch. However I think the messages are easily overlooked.

## Why should this be added

I believe that the current implementation is not loud enough, startup messages can be too easily overlooked. We just recently had someone on the discord, not realizing they had set caching to :null_store, despite having the newest SR version with the warnings.

## Changes

This PR:
* exits on faile sanity checks
* provides an config option as a way to prevent the app from exiting
```ruby
    StimulusReflex.configure do |config|
      config.exit_on_failed_sanity_checks = false
    end
```
* provides the info in the console on how to do so, when a check fails.
* gives a separate message if config.cache_store is set to :null_store
* always gets the version from `node_modules/stimulus_reflex/package.json` as `yarn.lock` may not contain a specific version, which leads to unclear feedback.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] Checks (StandardRB & Prettier-Standard) are passing

leastbad commented:
The `config.x` bit is strange to my eyes. Is that following a convention of some sort?

RolandStuder commented:
@leastbad yeah, the `config.x` looks strange, but it comes from rails and allows for nested configuration, instead of using `stimulus_reflex_ exit_for_failed_sanity_checks` (no dot) see https://guides.rubyonrails.org/configuring.html#custom-configuration.

I assume you can set up a custom config namespace somehow, but could not find out how to do it, my google fu failed me.If someone could please point me to a resource on how to do it, it would be appreciated.

marcoroth commented:
Would it make sense to introduce an initializer for configs like this?

Something like:

```ruby
# config/initializers/stimulus_reflex.rb

StimulusReflex.setup do |config|
  config.exit_for_failed_sanity_checks = true
end
```

leastbad commented:
I have been thinking about this PR a lot, especially since @marcoroth proposed an initializer.

While it's never been an explicit design goal to not have an initializer, I don't think it's an accident that we've gone this far without one. I do feel like there's a certain pride in being the sort of opinionated library that doesn't need an initializer, and if I'm honest, while I'm not opposed to introducing one, it does seem like the opportunity to turn off this functionality is a pretty weak reason to start down the initializer path.

You might remember that I was sad when we neutered the original proposed functionality as heavy-handed, opting for a server log message instead. What I know is that the love of my life will not read server log messages, even if they are flashing in bright red colors and have the answer to her questions right in front of her. She reads them like I read Linux boot sequence output - which is to say, under duress.

The facts, as I see them: SR will not work properly if the client+server versions are mismatched and lots of things force the Rails server process to bail at startup. I think it would be very different if it bailed any time after it was processing requests; we'd call that a bug. But I cannot think of a scenario in which I would consciously want to run mismatched SR versions even though it is pretty much guaranteed to explode.

So: my $0.02 is that we should just have the courage to exit, and scrap the part where it's an option or needs an initializer. We are helping them, and we know it works. If they want to run SR, they are going to want it to work.

RolandStuder commented:
I added a configuration class to avoid the `config.x.…` namespacing thingy. Currently it only works with a `configure do` but could of course be extended to allow access like `StimulusReflex.config. exit_for_failed_sanity_checks = false`

I added the configuration option only because I believe the default should be `exit` but for some reason we stepped back from this behavior in @excid3 PR, so I thought with an exit hatch it might be digestable.

excid3 commented:
If you're going to force an exit, it's good to have a way to disable that. I'm sure there are situations we haven't thought of and if the exit prevents someone from doing what they need, that would be bad.

excid3 commented:
Also, I vote for having it check for an ENV var as well as configs.

RolandStuder commented:
PR still needs an update. It fails when you do not call `.configure` 😒
 
I will also as discussed in the discord change the checks to always look in the package.json, rather than in yarn.lock as yarn lock can have stuff like `^3.3` something we did not anticipate.

excid3 commented:
Do you have an example of the lock line that does not match the regex?

marcoroth commented:
@excid3 the problem is: if you include `stimulus_reflex` with `>= 3.0` or `^3.3` in your `package.json` it will also write that in the `yarn.lock` and therefore it will not match with the current regex. 

We would need to match the line below the line we are currently matching the get the resolved version number. The file looks like this:

```js
// yarn.lock
[...]

"stimulus_reflex@>= 3.0":
  version "3.3.0"
  resolved "[...]"
  dependencies:
    "@rails/actioncable" ">= 6.0"
    cable_ready ">= 4.3.0"
    form-serialize ">= 0.7.2"
    stimulus ">= 1.1"

[...]
```

Or: we could just always read the version out of `node_modules/stimulus_reflex/package.json` because this file is present either way. And it would even make it yarn-independent.

excid3 commented:
Yeah but what I'm saying is should we even allow users to specify those
version numbers since it can allow mismatches?

On Tue, Oct 6, 2020 at 1:06 PM Marco Roth <notifications@github.com> wrote:

> @excid3 <https://github.com/excid3> the problem is: if you include
> stimulus_reflex with >= 3.0 or ^3.3 in your package.json it will also
> write that in the yarn.lock and therefore it will not match with the
> current regex.
>
> We would need to match the line below the line we are currently matching
> the get the resolved version number. The file looks like this:
>
> // yarn.lock[...]
> "stimulus_reflex@>= 3.0":
>   version "3.3.0"
>   resolved "[...]"
>   dependencies:
>     "@rails/actioncable" ">= 6.0"
>     cable_ready ">= 4.3.0"
>     form-serialize ">= 0.7.2"
>     stimulus ">= 1.1"
> [...]
>
> Or: we could just always read the version out of
> node_modules/stimulus_reflex/package.json because this file is present
> either way. And it would even make it yarn-independent.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hopsoft/stimulus_reflex/pull/318#issuecomment-704452349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAQMFORQQ5O55OON7M5LITSJNMCLANCNFSM4R6BKY4A>
> .
>
-- 
Chris Oliver
http://excid3.com | @excid3 <http://twitter.com/excid3> |
http://github.com/excid3

excid3 commented:
Whatever is the most robust. 👍

I was assuming that people would specify exact versions for added safety in version matching, but I'm sure that's not good enough for existing apps that didn't do that previously.

marcoroth commented:
Oh, I haven't seen your updated comment! 

Hmm. For our use case it would be nice if we would just allow exact versions. 
But since this is a feature of both bundler and yarn it seems a little bit off to just allow exact versions.

leastbad commented:
I think we're tripping if we think we can demand people do the right thing.

If that was something we could do, we wouldn't be discussing how to help people make sure their package versions are in sync.

RolandStuder commented:
Ok, PR is updated:
* now to call `StimulusReflex.configure` is optional.
* the package version always is taken from `node_modules/stimulus_reflex/package.json `

RolandStuder commented:
@excid3 I don't understand the need for an ENV variable. Do we really need that? If you want to be flexible you can use a custom ENV variable in you initializer. What am I missing? Is it just a question of convenience?

RolandStuder commented on lib/stimulus_reflex/sanity_checker.rb:
```diff
@@ -53,14 +62,11 @@ def gem_version
   def javascript_package_version
-    return @_js_version if defined?(@_js_version)
-    @_js_version = find_javascript_package_version
+    @_js_version ||= find_javascript_package_version
```
Do we actually need the memoization? We could save a method if we don't memoize :-)

leastbad replied:
It's actually used more than the gem version equivalent.

leastbad commented:
Hey gang, I just wanted to make sure that your hard work is coordinated with a different set of changes being proposed in https://github.com/hopsoft/stimulus_reflex/pull/322

I made some noise pushing back against an initializer earlier in the week; I guess I'm over that now? 😁 

Anyhow, one of these PRs will get merged first, and it could be 322 at this rate.

RolandStuder commented:
@leastbad this PR is now ready and uses the config from #339, if had to force push as I changed the commit history, but this way the git history will be cleaner.
